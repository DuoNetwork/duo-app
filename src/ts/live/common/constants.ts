export * from '../../../../../duo-admin/src/constants';

export const AC_CTD_STATES = 'custodianStates';
export const AC_CTD_PRICES = 'custodianPrices';
export const AC_ACCOUNT = 'account';
export const AC_NETWORK = 'network';
export const AC_BALANCES = 'balances';
export const AC_ADDRESSES = 'addresses';
export const AC_DNM_STATUS = 'dynamoStatus';
export const AC_DMN_HOURLY = 'dynamoHourly';
export const AC_DMN_MINUTELY = 'dynamoMinutely';
export const AC_DMN_PRICE = 'dynamoPrice';
export const AC_TOTAL_SUPPLY = 'totalSupply';
export const AC_CONVERSION = 'conversion';
export const AC_REFRESH = 'refresh';

export const ETH_MAINNET_ID = 1;
export const ETH_KOVAN_ID = 42;

export const CTD_INCEPTION = 'Inception';
export const CTD_TRADING = 'Trading'
export const CTD_PRERESET = 'PreReset';
export const CTD_UP_RESET = 'UpwardReset';
export const CTD_DOWN_RESET = 'DownwardReset';
export const CTD_PERIOD_RESET = 'PeriodicReset';
export const CTD_LOADING = 'Loading';
