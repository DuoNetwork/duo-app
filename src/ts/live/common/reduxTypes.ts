import { IBalances, ICustodianPrice, ICustodianStates } from './types';

export interface IState {
	contract: IContractState;
}

export interface IContractState {
	custodianStates: ICustodianStates;
	custodianPrices: {
		reset: ICustodianPrice;
		last: ICustodianPrice;
		// first: ICustodianPrice;
		// second: ICustodianPrice;
	};
	balances: IBalances;
}

export type Action = IBaseAction | IBooleanAction | IStringAction | IObjectAction | INumberAction;

export type Dispatch = (action: Action | ThunkAction | PromiseAction) => any;

export interface IBaseAction {
	type: string;
}

export interface IBooleanAction {
	type: string;
	value: boolean;
}

export interface IStringAction {
	type: string;
	value: string;
}

export interface IObjectAction {
	type: string;
	value: object;
}

export interface INumberAction {
	type: string;
	value: number;
}

export type ThunkAction = (dispatch: Dispatch) => any;
export type PromiseAction = Promise<Action>;
