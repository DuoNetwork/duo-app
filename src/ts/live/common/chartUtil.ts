import moment from 'moment';
import * as CST from '../common/constants';
import { IAcceptedPrice, ICustodianPrice, ICustodianStates, IPriceBar, ISourceData } from './types';

class ChartUtil {
	public interpolate(sourceData: IPriceBar[], isHourly: boolean): IPriceBar[] {
		if (!sourceData.length) return [];
		sourceData.sort((a, b) => a.timestamp - b.timestamp);
		const newSourceData: IPriceBar[] = [];
		const timeStep = isHourly ? 3.6e6 : 60000;
		const source = sourceData[0].source;
		for (let i = 0; i < sourceData.length - 1; i++) {
			newSourceData.push(sourceData[i]);
			const timeIndex = sourceData[i].timestamp;
			const nextTimeIndex = sourceData[i + 1].timestamp;
			const close = sourceData[i].close;
			let newTimeStamp = timeIndex + timeStep;
			while (newTimeStamp < nextTimeIndex) {
				newSourceData.push({
					source: source,
					date: moment(newTimeStamp).format('YYYY-MM-DD'),
					hour: moment(newTimeStamp).format('HH'),
					minute: isHourly ? 0 : Number(moment(newTimeStamp).format('mm')),
					open: close,
					high: close,
					low: close,
					close: close,
					volume: 0,
					timestamp: newTimeStamp
				});
				newTimeStamp += timeStep;
			}
		}
		newSourceData.push(sourceData[sourceData.length - 1]);

		return newSourceData;
	}

	public reset(
		price: IAcceptedPrice[],
		limitUp: number,
		limitDown: number,
		limitPeriod: number
	): IAcceptedPrice[] {
		return price
			.filter(p => p.navB >= limitUp || p.navB <= limitDown || p.navA >= limitPeriod)
			.map(p => ({
				price: p.price,
				navA: 1,
				navB: 1,
				timestamp: p.timestamp + 717
			}));
	}

	public mergeReset(prices: IAcceptedPrice[], resets: IAcceptedPrice[]): IAcceptedPrice[] {
		const all = [...prices, ...resets];
		all.sort((a, b) => a.timestamp - b.timestamp);
		return all;
	}

	public mergeLastToPriceBar(
		pricebars: ISourceData<IPriceBar[]>,
		last: ISourceData<ICustodianPrice>,
		isHourly: boolean
	) {
		CST.EXCHANGES.forEach(src => {
			const pricebar: IPriceBar[] = pricebars[src.toLowerCase()];
			if (!pricebar || !pricebar.length) return;
			const sourceLast: ICustodianPrice = last[src.toLowerCase()];
			const lastBar = pricebar[pricebar.length - 1];
			const dateObj = moment.utc(sourceLast.timestamp);
			if (isHourly && lastBar.date + ' ' + lastBar.hour !== dateObj.format('YYYY-MM-DD HH'))
				pricebar.push({
					source: lastBar.source,
					date: dateObj.format('YYYY-MM-DD'),
					hour: dateObj.format('HH'),
					minute: 0,
					open: lastBar.close,
					high: Math.max(lastBar.close, sourceLast.price),
					low: Math.min(lastBar.close, sourceLast.price),
					close: sourceLast.price,
					volume: 0,
					timestamp: sourceLast.timestamp
				});
			else if (
				!isHourly &&
				lastBar.date + ' ' + lastBar.hour + ' ' + lastBar.minute !==
					dateObj.format('YYYY-MM-DD HH m')
			)
				pricebar.push({
					source: lastBar.source,
					date: dateObj.format('YYYY-MM-DD'),
					hour: dateObj.format('HH'),
					minute: Number(dateObj.format('mm')),
					open: lastBar.close,
					high: Math.max(lastBar.close, sourceLast.price),
					low: Math.min(lastBar.close, sourceLast.price),
					close: sourceLast.price,
					volume: 0,
					timestamp: sourceLast.timestamp
				});
			else {
				lastBar.high = Math.max(lastBar.high, sourceLast.price);
				lastBar.low = Math.min(lastBar.low, sourceLast.price);
				lastBar.close = sourceLast.price;
				lastBar.timestamp = sourceLast.timestamp;
			}
		});
	}

	public mergeLastToPrice(
		prices: IAcceptedPrice[],
		states: ICustodianStates,
		last: ICustodianPrice
	) {
		if (!prices.length) return;
		const lastPrice = prices[prices.length - 1];
		if (lastPrice.timestamp < last.timestamp)
			prices.push({
				price: last.price,
				navA: states.navA,
				navB: states.navB,
				timestamp: last.timestamp
			});
	}
}

const chartUtil = new ChartUtil();
export default chartUtil;
