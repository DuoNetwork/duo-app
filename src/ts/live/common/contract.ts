import infura from '../../../../../duo-admin/src/keys/infura.json';
import ContractUtil from '../../../../../duo-contract-util/src/contractUtil';
import * as CST from './constants';

const provider =
	(__KOVAN__ ? CST.PROVIDER_INFURA_KOVAN : CST.PROVIDER_INFURA_MAIN) + '/' + infura.token;
const contract = new ContractUtil(window, '', provider, !__KOVAN__);
export default contract;
