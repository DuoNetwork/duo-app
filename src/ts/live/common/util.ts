import * as d3 from 'd3';
import moment, { DurationInputArg2 } from 'moment';
import * as CST from './constants';
import { ICustodianPrice, IPriceStatus, ISourceData, IStatus } from './types';

class Util {
	public convertUpdateTime(timestamp: number): string {
		const diff = this.getUTCNowTimestamp() - timestamp;
		if (diff < 60000) return 'Just Now';
		else if (diff < 3600000) return Math.floor(diff / 60000) + ' Minutes Ago';
		else if (diff < 86400000) return Math.floor(diff / 3600000) + ' Hours Ago';
		else if (diff < 2592000000) return Math.floor(diff / 86400000) + ' Days Ago';
		else return 'Long Time Ago';
	}

	public getUTCNowTimestamp() {
		return moment().valueOf();
	}

	public calculateNav(
		price: number,
		time: number,
		resetPrice: number,
		resetTime: number,
		alpha: number,
		beta: number,
		period: number,
		coupon: number
	) {
		const navParent = (price / resetPrice / beta) * (1 + alpha);

		const navA = 1 + Math.floor((time - resetTime) / 1000 / period) * coupon;
		const navAAdj = navA * alpha;
		if (navParent <= navAAdj) return [navParent / alpha, 0];
		else return [navA, navParent - navAAdj];
	}

	public getDates(length: number, step: number, stepSize: DurationInputArg2, format: string) {
		const dates: string[] = [];
		const date = moment.utc();
		for (let i = 0; i < length; i++) {
			dates.push(date.format(format));
			date.subtract(step, stepSize);
		}
		dates.sort((a, b) => a.localeCompare(b));

		return dates;
	}

	public round(num: number) {
		return +(Math.floor((num + 'e+8') as any) + 'e-8');
	}

	public range(start: number, end: number) {
		const subArray = Array.apply(null, { length: end }).map(Number.call, Number);
		return subArray.slice(start, end);
	}

	public formatBalance(num: number) {
		if (Math.abs(num) < 1e-8) return '0.000';
		return d3
			.format(Math.abs(num) > 1 ? ',.4s' : ',.4n')(num)
			.toUpperCase()
			.replace(/G/g, 'B');
	}

	public formatNumber(num: number) {
		if (Math.abs(num) < 1e-8) return '0.000';
		if (Math.abs(num) < 1) return d3.format(',.4n')(num);
		if (Math.abs(num) < 100000) return d3.format(',.2f')(num);
		return d3
			.format(',.4s')(num)
			.toUpperCase()
			.replace(/G/g, 'B');
	}

	public getCurrentPath(location: any) {
		return location.pathname === '/' ? location.pathname + 'beethoven' : location.pathname;
	}

	public getLastPriceFromStatus(status: IStatus[]): ISourceData<ICustodianPrice> {
		// const bitfinex: ICustodianPrice = {
		// 	address: CST.DUMMY_ADDR,
		// 	price: 0,
		// 	timestamp: 0
		// };
		const kraken: ICustodianPrice = {
			address: CST.DUMMY_ADDR,
			price: 0,
			timestamp: 0
		};
		const gemini: ICustodianPrice = {
			address: CST.DUMMY_ADDR,
			price: 0,
			timestamp: 0
		};
		const gdax: ICustodianPrice = {
			address: CST.DUMMY_ADDR,
			price: 0,
			timestamp: 0
		};
		status.forEach(s => {
			// if (s.process === 'PRICE_AWS_PUBLIC_BITFINEX') {
			// 	bitfinex.price = (s as IPriceStatus).price;
			// 	bitfinex.timestamp = (s as IPriceStatus).timestamp;
			// } else
			if (s.process === 'TRADE_AWS_PUBLIC_GEMINI') {
				gemini.price = (s as IPriceStatus).price;
				gemini.timestamp = (s as IPriceStatus).timestamp;
			} else if (s.process === 'TRADE_AWS_PUBLIC_KRAKEN') {
				kraken.price = (s as IPriceStatus).price;
				kraken.timestamp = (s as IPriceStatus).timestamp;
			} else if (s.process === 'TRADE_AWS_PUBLIC_GDAX') {
				gdax.price = (s as IPriceStatus).price;
				gdax.timestamp = (s as IPriceStatus).timestamp;
			}
		});

		return {
			// bitfinex,
			kraken,
			gemini,
			gdax
		};
	}
}

const util = new Util();
export default util;
