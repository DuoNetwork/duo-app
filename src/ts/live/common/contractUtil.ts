import Web3 from 'web3';
import { Contract } from 'web3/types';
import infura from '../../../../../duo-admin/src/keys/infura.json';
import custodianAbi from '../../../../../duo-admin/src/static/Custodian.json';
import duoAbi from '../../../../../duo-admin/src/static/DUO.json';
import { IAddresses, IBalances, ICustodianPrices, ICustodianStates } from '../common/types';
import * as CST from './constants';
//import util from './util';

class ContractUtil {
	private web3: Web3;
	private duo: Contract;
	private custodian: Contract;
	private isReadOnly: boolean;
	public readonly custodianAddr: string;
	private readonly duoContractAddr: string;

	constructor() {
		if (typeof (window as any).web3 !== 'undefined') {
			this.isReadOnly = false;
			this.web3 = new Web3((window as any).web3.currentProvider);
		} else {
			this.web3 = new Web3(
				new Web3.providers.HttpProvider(
					(__KOVAN__ ? CST.PROVIDER_INFURA_KOVAN : CST.PROVIDER_INFURA_MAIN) +
						'/' +
						infura.token
				)
			);
			this.isReadOnly = true;
		}
		this.custodianAddr = __KOVAN__ ? CST.CUSTODIAN_ADDR_KOVAN : CST.CUSTODIAN_ADDR_MAIN;
		this.custodian = new this.web3.eth.Contract(custodianAbi.abi, this.custodianAddr);
		this.duoContractAddr = __KOVAN__ ? CST.DUO_CONTRACT_ADDR_KOVAN : CST.DUO_CONTRACT_ADDR_MAIN;
		this.duo = new this.web3.eth.Contract(duoAbi.abi, this.duoContractAddr);
	}

	public onWeb3AccountUpdate(onUpdate: (addr: string, network: number) => any) {
		const store = (this.web3.currentProvider as any).publicConfigStore;
		if (store)
			store.on('update', () => {
				onUpdate(
					store.getState().selectedAddress || '',
					Number(store.getState().networkVersion || '')
				);
			});
	}

	public convertCustodianState(rawState: string) {
		switch (rawState) {
			case CST.STATE_INCEPTION:
				return CST.CTD_INCEPTION;
			case CST.STATE_TRADING:
				return CST.CTD_TRADING;
			case CST.STATE_PRERESET:
				return CST.CTD_PRERESET;
			case CST.STATE_UP_RESET:
				return CST.CTD_UP_RESET;
			case CST.STATE_DOWN_RESET:
				return CST.CTD_DOWN_RESET;
			case CST.STATE_PERIOD_RESET:
				return CST.CTD_PERIOD_RESET;
			default:
				return CST.CTD_LOADING;
		}
	}

	public async getSystemStates(): Promise<ICustodianStates> {
		const [states, duoBalance] = await Promise.all([
			this.custodian.methods.getSystemStates().call(),
			this.getDuoBalance(this.custodianAddr)
		]);
		return {
			state: this.convertCustodianState(states[0].valueOf()),
			navA: this.fromWei(states[1]),
			navB: this.fromWei(states[2]),
			totalSupplyA: this.fromWei(states[3]),
			totalSupplyB: this.fromWei(states[4]),
			ethBalance: this.fromWei(states[5]),
			alpha: states[6].valueOf() / 10000,
			beta: this.fromWei(states[7]),
			feeAccumulated: this.fromWei(states[8]),
			periodCoupon: this.fromWei(states[9]),
			limitPeriodic: this.fromWei(states[10]),
			limitUpper: this.fromWei(states[11]),
			limitLower: this.fromWei(states[12]),
			commissionRate: states[13] / 10000,
			period: Number(states[14].valueOf()),
			iterationGasThreshold: Number(states[15].valueOf()),
			ethDuoFeeRatio: Number(states[16].valueOf()),
			preResetWaitingBlocks: Number(states[17].valueOf()),
			priceTol: Number(states[18].valueOf() / 10000),
			priceFeedTol: Number(states[19].valueOf() / 10000),
			priceFeedTimeTol: Number(states[20].valueOf()),
			priceUpdateCoolDown: Number(states[21].valueOf()),
			numOfPrices: Number(states[22].valueOf()),
			nextResetAddrIndex: Number(states[23].valueOf()),
			lastAdminTime: Number(states[24].valueOf()),
			adminCoolDown: Number(states[25]),
			usersLength: Number(states[26].valueOf()),
			addrPoolLength: Number(states[27].valueOf()),
			duoBalance: duoBalance
		};
	}

	public async getSystemAddresses(): Promise<IAddresses> {
		const addr: string[] = await this.custodian.methods.getSystemAddresses().call();
		const balances = await Promise.all(addr.map(a => this.getEthBalance(a)));
		return {
			operator: {
				address: addr[0],
				balance: balances[0]
			},
			feeCollector: {
				address: addr[1],
				balance: balances[1]
			},
			priceFeed1: {
				address: addr[2],
				balance: balances[2]
			},
			priceFeed2: {
				address: addr[3],
				balance: balances[3]
			},
			priceFeed3: {
				address: addr[4],
				balance: balances[4]
			},
			poolManager: {
				address: addr[5],
				balance: balances[5]
			}
		};
	}

	public async getSystemPrices(): Promise<ICustodianPrices> {
		const prices = await this.custodian.methods.getSystemPrices().call();
		const custodianPrices = [0, 1, 2, 3].map(i => ({
			address: prices[i * 3].valueOf(),
			price: this.fromWei(prices[1 + i * 3]),
			timestamp: prices[2 + i * 3].valueOf() * 1000
		}));

		return {
			first: custodianPrices[0],
			second: custodianPrices[1],
			reset: custodianPrices[2],
			last: custodianPrices[3]
		};
	}

	public async getBalances(address: string): Promise<IBalances> {
		if (!address)
			return {
				eth: 0,
				duo: 0,
				allowance: 0,
				tokenA: 0,
				tokenB: 0
			};

		const balances = await Promise.all([
			this.getEthBalance(address),
			this.getDuoBalance(address),
			this.getDuoAllowance(address),
			this.getTokenBalance(address, true),
			this.getTokenBalance(address, false)
		]);

		return {
			eth: balances[0],
			duo: balances[1],
			allowance: balances[2],
			tokenA: balances[3],
			tokenB: balances[4]
		};
	}

	public getUserAddress(index: number) {
		return this.custodian.methods.users(index).call();
	}

	public getPoolAddress(index: number) {
		return this.custodian.methods.addrPool(index).call();
	}

	public async getGasPrice(): Promise<number> {
		return await this.web3.eth.getGasPrice();
	}

	public async getCurrentAddress(): Promise<string> {
		return (await this.web3.eth.getAccounts())[0];
	}

	public getCurrentNetwork(): Promise<number> {
		return this.web3.eth.net.getId();
	}

	public async getEthBalance(address: string): Promise<number> {
		return this.fromWei(await this.web3.eth.getBalance(address));
	}

	private async getDuoBalance(address: string): Promise<number> {
		return this.fromWei(await this.duo.methods.balanceOf(address).call());
	}

	private async getDuoAllowance(address: string): Promise<number> {
		return this.fromWei(await this.duo.methods.allowance(address, this.custodianAddr).call());
	}

	private async getTokenBalance(address: string, isA: boolean): Promise<number> {
		return this.fromWei(await this.custodian.methods.balanceOf(isA ? 0 : 1, address).call());
	}

	public fromWei(value: string | number) {
		return Number(this.web3.utils.fromWei(value, 'ether'));
	}

	public toWei(value: string | number) {
		return this.web3.utils.toWei(value + '', 'ether');
	}

	public checkAddress(addr: string) {
		if (!addr.startsWith('0x') || addr.length !== 42) return false;
		return this.web3.utils.checkAddressChecksum(addr);
	}

	public create(
		address: string,
		value: number,
		payFeeInEth: boolean,
		onTxHash: (hash: string) => any
	) {
		if (this.isReadOnly) return Promise.reject('Read Only Mode');

		return this.custodian.methods
			.create(payFeeInEth)
			.send({
				from: address,
				value: this.toWei(value)
			})
			.on('transactionHash', onTxHash);
	}

	public redeem(
		address: string,
		amtA: number,
		amtB: number,
		payFeeInEth: boolean,
		onTxHash: (hash: string) => any
	) {
		if (this.isReadOnly) return Promise.reject('Read Only Mode');

		return this.custodian.methods
			.redeem(this.toWei(amtA), this.toWei(amtB), payFeeInEth)
			.send({
				from: address
			})
			.on('transactionHash', onTxHash);
	}

	public duoApprove(address: string, spender: string, value: number) {
		if (this.isReadOnly) return Promise.reject('Read Only Mode');

		return this.duo.methods.approve(spender, this.toWei(value)).send({
			from: address
		});
	}

	public duoTransfer(address: string, to: string, value: number) {
		if (this.isReadOnly) return Promise.reject('Read Only Mode');

		return this.duo.methods.transfer(to, this.toWei(value)).send({
			from: address
		});
	}

	public approve(address: string, spender: string, value: number, isA: boolean) {
		if (this.isReadOnly) return Promise.reject('Read Only Mode');

		return this.custodian.methods.approve(isA ? 0 : 1, spender, this.toWei(value)).send({
			from: address
		});
	}

	public transfer(address: string, to: string, value: number, isA: boolean) {
		if (this.isReadOnly) return Promise.reject('Read Only Mode');

		// dummy from address
		return this.custodian.methods.transfer(isA ? 0 : 1, address, to, this.toWei(value)).send({
			from: address
		});
	}

	public collectFee(address: string, amount: number) {
		if (this.isReadOnly) return Promise.reject('Read Only Mode');
		return this.custodian.methods.collectFee(this.toWei(amount)).send({
			from: address
		});
	}

	public setValue(address: string, index: number, newValue: number) {
		if (this.isReadOnly) return Promise.reject('Read Only Mode');
		return this.custodian.methods.setValue(index, newValue).send({
			from: address
		});
	}

	public addAddress(address: string, addr1: string, addr2: string) {
		if (this.isReadOnly) return Promise.reject('Read Only Mode');
		return this.custodian.methods.addAddress(addr1, addr2).send({
			from: address
		});
	}

	public removeAddress(address: string, addr: string) {
		if (this.isReadOnly) return Promise.reject('Read Only Mode');
		return this.custodian.methods.addAddress(addr).send({
			from: address
		});
	}

	public updateAddress(address: string, currentRole: string) {
		if (this.isReadOnly) return Promise.reject('Read Only Mode');
		return this.custodian.methods.addAddress(currentRole).send({
			from: address
		});
	}
}

const contractUtil = new ContractUtil();
export default contractUtil;
