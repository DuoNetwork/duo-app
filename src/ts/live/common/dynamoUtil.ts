import AWS from 'aws-sdk';
import { ScanInput, ScanOutput } from 'aws-sdk/clients/dynamodb';
import devConfig from '../../keys/aws.ui.dev.json';
import liveConfig from '../../keys/aws.ui.live.json';
import { IPriceStatus, IStatus } from '../common/types';
import * as CST from './constants';

export class DynamoUtil {
	private ddb: AWS.DynamoDB;
	constructor() {
		// this.role = role;
		AWS.config.update(__DEV__ ? devConfig : liveConfig);
		this.ddb = new AWS.DynamoDB({ apiVersion: CST.AWS_DYNAMO_API_VERSION });
	}

	public scanData(params: ScanInput): Promise<ScanOutput> {
		return new Promise((resolve, reject) => {
			this.ddb.scan(params, (err, data) => {
				if (err) reject(err);
				else resolve(data);
			});
		});
	}

	public async scanStatus() {
		return this.convertStatus(
			await this.scanData({
				TableName: __DEV__ ? CST.DB_AWS_STATUS_DEV : CST.DB_AWS_STATUS_LIVE
			})
		);
	}

	public convertStatus(status: ScanOutput): IStatus[] {
		if (!status.Items || !status.Items.length) return [];

		const output =  status.Items.map(d => {
			const process = d[CST.DB_ST_PROCESS].S || '';
			const timestamp = Number(d[CST.DB_ST_TS].N);
			if (process.startsWith('PRICE'))
				return {
					process: process,
					timestamp: timestamp,
					price: Number(d[CST.DB_TX_PRICE].N),
					volume: Number(d[CST.DB_TX_AMOUNT].N)
				} as IPriceStatus;
			else if (process.startsWith('FEED'))
				return {
					process: process,
					timestamp: timestamp,
					price: Number(d[CST.DB_HISTORY_PRICE].N),
					volume: Number(d[CST.DB_HISTORY_VOLUME].N)
				} as IPriceStatus;
			else
				return {
					process: process,
					timestamp: timestamp
				};
		});
		output.sort((a, b) => b.timestamp - a.timestamp);
		output.sort((a, b) => a.process.localeCompare(b.process));

		return output;
	}
}

const dynamoUtil = new DynamoUtil();
export default dynamoUtil;
