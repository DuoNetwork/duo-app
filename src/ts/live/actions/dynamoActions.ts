import moment from 'moment';
import chartUtil from '../common/chartUtil';
import * as CST from '../common/constants';
import dynamoUtil from '../common/dynamoUtil';
import * as reduxTypes from '../common/reduxTypes';
import { IPriceBars } from '../common/types';

export function statusUpdate(status: object): reduxTypes.Action {
	return {
		type: CST.AC_DNM_STATUS,
		value: status
	};
}

export function scanStatus(): reduxTypes.ThunkAction {
	return async dispatch => {
		const states = await dynamoUtil.scanStatus();
		dispatch(statusUpdate(states));
	};
}

export function hourlyUpdate(hourly: IPriceBars): reduxTypes.Action {
	return {
		type: CST.AC_DMN_HOURLY,
		value: hourly
	};
}

export function fetchHourly(): reduxTypes.ThunkAction {
	return async dispatch => {
		const dates: string[] = [];
		const date = moment.utc();
		for (let i = 0; i < 7; i++) {
			dates.push(date.format('YYYY-MM-DD'))
			date.subtract(1, 'day');
		}
		const promistList = CST.EXCHANGES.map(src => dynamoUtil.queryHourlyOHLC(src, dates));
		const results = await Promise.all(promistList);
		const hourly: IPriceBars = {
			bitfinex: [],
			gemini: [],
			kraken: [],
			gdax: []
		};
		results.forEach((r, i) => (hourly[CST.EXCHANGES[i].toLowerCase()] = chartUtil.interpolate(r, true)));
		dispatch(hourlyUpdate(hourly));
	};
}

export function minutelyUpdate(minutely: IPriceBars): reduxTypes.Action {
	return {
		type: CST.AC_DMN_MINUTELY,
		value: minutely
	};
}

export function fetchMinutely(): reduxTypes.ThunkAction {
	return async dispatch => {
		const dates: string[] = [];
		const date = moment.utc();
		for (let i = 0; i < 2; i++) {
			dates.push(date.format('YYYY-MM-DD-HH'))
			date.subtract(1, 'hour');
		}
		const promistList = CST.EXCHANGES.map(src => dynamoUtil.queryMinutelyOHLC(src, dates));
		const results = await Promise.all(promistList);
		const minutely: IPriceBars = {
			bitfinex: [],
			gemini: [],
			kraken: [],
			gdax: []
		};
		results.forEach((r, i) => (minutely[CST.EXCHANGES[i].toLowerCase()] = chartUtil.interpolate(r, false)));
		dispatch(minutelyUpdate(minutely));
	};
}
