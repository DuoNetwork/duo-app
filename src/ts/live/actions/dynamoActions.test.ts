import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import dynamoUtil from '../../../../../duo-admin/src/utils/dynamoUtil';
import * as CST from '../common/constants';
import util from '../common/util';
import * as dynamoActions from './dynamoActions';

const mockStore = configureMockStore([thunk]);

describe('actions', () => {
	test('statusUpdate', () => {
		expect(dynamoActions.statusUpdate({ test: 'test' } as any)).toMatchSnapshot();
	});

	test('scanStatus', () => {
		const store: any = mockStore({});
		dynamoUtil.scanStatus = jest.fn(() =>
			Promise.resolve({
				test: 'test'
			})
		);
		store.dispatch(dynamoActions.scanStatus());
		return new Promise(resolve =>
			setTimeout(() => {
				expect(store.getActions()).toMatchSnapshot();
				resolve();
			}, 0)
		);
	});

	test('pricesUpdate', () => {
		expect(dynamoActions.pricesUpdate(['test'] as any)).toMatchSnapshot();
	});

	test('fetchPrices 60', () => {
		util.getUTCNowTimestamp = jest.fn(() => 1234567890);
		const store: any = mockStore({
			ui: {
				period: 60,
				source: 'test'
			}
		});
		dynamoUtil.getPrices = jest.fn(
			(src: string, period: number, start: number, end: number, pair: string) =>
				Promise.resolve([src, period, start, end, pair])
		);
		store.dispatch(dynamoActions.fetchPrices());
		return new Promise(resolve =>
			setTimeout(() => {
				expect(store.getActions()).toMatchSnapshot();
				resolve();
			}, 0)
		);
	});

	test('fetchPrices 5', () => {
		util.getUTCNowTimestamp = jest.fn(() => 1234567890);
		const store: any = mockStore({
			ui: {
				period: 5,
				source: 'test'
			}
		});
		dynamoUtil.getPrices = jest.fn(
			(src: string, period: number, start: number, end: number, pair: string) =>
				Promise.resolve([src, period, start, end, pair])
		);
		store.dispatch(dynamoActions.fetchPrices());
		return new Promise(resolve =>
			setTimeout(() => {
				expect(store.getActions()).toMatchSnapshot();
				resolve();
			}, 0)
		);
	})

	test('acceptedPricesUpdate', () => {
		expect(dynamoActions.acceptedPricesUpdate(['test'] as any)).toMatchSnapshot();
	});

	test('fetchAcceptedPrices', () => {
		const store: any = mockStore({
			contract: { states: { limitUpper: 2, limitLower: 0.25, limitPeriodic: 1.035 } }
		});
		dynamoUtil.queryAcceptPriceEvent = jest.fn(() => Promise.resolve(['test']));
		store.dispatch(dynamoActions.fetchAcceptedPrices(CST.DUMMY_ADDR));
		return new Promise(resolve =>
			setTimeout(() => {
				expect(store.getActions()).toMatchSnapshot();
				resolve();
			}, 1000)
		);
	});

	test('conversionsUpdate', () => {
		expect(dynamoActions.conversionsUpdate(['test'] as any)).toMatchSnapshot();
	});

	test('fetchConversions', () => {
		const store: any = mockStore({ contract: { account: CST.DUMMY_ADDR } });
		util.getDates = jest.fn(() => ['1970-01-15']);
		dynamoUtil.queryConversionEvent = jest.fn(() =>
			Promise.resolve([
				{
					transactionHash: 'aaa'
				}
			])
		);
		dynamoUtil.queryUIConversionEvent = jest.fn(() =>
			Promise.resolve([
				{
					transactionHash: 'aaa',
					timestamp: 1234567890
				},
				{
					transactionHash: 'bbb',
					timestamp: 1234567890
				},
				{
					transactionHash: 'ccc',
					timestamp: 0
				}
			])
		);
		dynamoUtil.deleteUIConversionEvent = jest.fn(() => Promise.resolve());
		store.dispatch(dynamoActions.fetchConversions(CST.DUMMY_ADDR));
		return new Promise(resolve =>
			setTimeout(() => {
				expect(store.getActions()).toMatchSnapshot();
				expect(
					(dynamoUtil.deleteUIConversionEvent as jest.Mock<Promise<void>>).mock.calls
						.length
				).toBe(2);
				expect(
					(dynamoUtil.deleteUIConversionEvent as jest.Mock<Promise<void>>).mock
						.calls[0][1]
				).toMatchSnapshot();
				expect(
					(dynamoUtil.deleteUIConversionEvent as jest.Mock<Promise<void>>).mock
						.calls[1][1]
				).toMatchSnapshot();
				resolve();
			}, 1000)
		);
	});
});
