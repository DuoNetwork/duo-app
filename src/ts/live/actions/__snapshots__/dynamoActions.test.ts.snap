// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actions conversionUpdate 1`] = `
Object {
  "type": "conversion",
  "value": Array [
    "test",
  ],
}
`;

exports[`actions fetchConversion 1`] = `
Array [
  Object {
    "type": "conversion",
    "value": Array [
      "test",
    ],
  },
]
`;

exports[`actions fetchHourly 1`] = `
Array [
  Object {
    "type": "dynamoHourly",
    "value": Object {
      "bitfinex": Object {
        "test": "test",
      },
      "gdax": Object {
        "test": "test",
      },
      "gemini": Object {
        "test": "test",
      },
      "kraken": Object {
        "test": "test",
      },
    },
  },
]
`;

exports[`actions fetchMinutely 1`] = `
Array [
  Object {
    "type": "dynamoMinutely",
    "value": Object {
      "bitfinex": Object {
        "test": "test",
      },
      "gdax": Object {
        "test": "test",
      },
      "gemini": Object {
        "test": "test",
      },
      "kraken": Object {
        "test": "test",
      },
    },
  },
]
`;

exports[`actions fetchPrice 1`] = `
Array [
  Object {
    "type": "dynamoPrice",
    "value": Array [
      "test",
    ],
  },
]
`;

exports[`actions fetchTotalSupply 1`] = `
Array [
  Object {
    "type": "totalSupply",
    "value": Array [
      "test",
    ],
  },
]
`;

exports[`actions hourlyUpdate 1`] = `
Object {
  "type": "dynamoHourly",
  "value": Object {
    "test": "test",
  },
}
`;

exports[`actions minutelyUpdate 1`] = `
Object {
  "type": "dynamoMinutely",
  "value": Object {
    "test": "test",
  },
}
`;

exports[`actions priceUpdate 1`] = `
Object {
  "type": "dynamoPrice",
  "value": Array [
    "test",
  ],
}
`;

exports[`actions scanStatus 1`] = `
Array [
  Object {
    "type": "dynamoStatus",
    "value": Object {
      "test": "test",
    },
  },
]
`;

exports[`actions statusUpdate 1`] = `
Object {
  "type": "dynamoStatus",
  "value": Object {
    "test": "test",
  },
}
`;

exports[`actions totalSupplyUpdate 1`] = `
Object {
  "type": "totalSupply",
  "value": Array [
    "test",
  ],
}
`;
